K-means: support for array input

orhankislal opened the pull request:
JIRA: MADLIB-1018

Adds support for array input as data points. The function collates
the columns into a column in a temp table.

decibel commented on src/ports/postgres/modules/kmeans/kmeans.py_in:
```diff
@@ -34,6 +37,25 @@ def kmeans_validate_src(schema_madlib, rel_source, **kwargs):
 
 # ----------------------------------------------------------------------
 
+def kmeans_validate_expr(schema_madlib, rel_source, expr_point, **kwargs):
```
This function could really use some comments or a docstring about what it's supposed to be doing.

decibel commented on src/ports/postgres/modules/kmeans/kmeans.py_in:
```diff
@@ -34,6 +37,25 @@ def kmeans_validate_src(schema_madlib, rel_source, **kwargs):
+def kmeans_validate_expr(schema_madlib, rel_source, expr_point, **kwargs):
+    if not columns_exist_in_table(rel_source, [expr_point]):
+
+        p = re.compile('[Aa][Rr][Rr][Aa][Yy]\s*\[\s*[\s*\w|,\s*]+\s*\]')
```
A comment on what the regex is doing would be very helpful. The array part is pretty obvious (though, couldn't that just be done with an insensitive regex??), but the rest is more difficult to follow.

decibel commented on src/ports/postgres/modules/kmeans/kmeans.py_in:
```diff
@@ -34,6 +37,25 @@ def kmeans_validate_src(schema_madlib, rel_source, **kwargs):
 # ----------------------------------------------------------------------
 
+def kmeans_validate_expr(schema_madlib, rel_source, expr_point, **kwargs):
+    if not columns_exist_in_table(rel_source, [expr_point]):
```
I think it'd be much clearer to make this if columns_exist... and return immediately if true.

decibel commented on src/ports/postgres/modules/kmeans/kmeans.py_in:
```diff
@@ -34,6 +37,25 @@ def kmeans_validate_src(schema_madlib, rel_source, **kwargs):
+                """kmeans error: {expr_point} does not exist in {rel_source}!
+                """.format(**locals()))
+    return rel_source, False
+
```
It would be kinda nice to pre-compile the re by sticking a `kmeans_validate_expr.p = re.compile(...)` down here, as per http://stackoverflow.com/questions/279561/what-is-the-python-equivalent-of-static-variables-inside-a-function. Others might well have different ideas about that, though.

decibel commented on src/ports/postgres/modules/kmeans/kmeans.py_in:
```diff
@@ -34,6 +37,25 @@ def kmeans_validate_src(schema_madlib, rel_source, **kwargs):
+        if p.match(expr_point.strip()):
+            view_name = unique_string('km_view')
+
+            plpy.execute(""" CREATE VIEW {view_name} AS
```
Would this be better as a TEMP view? Or do all the callers create permanent tables?

decibel commented on src/ports/postgres/modules/kmeans/kmeans.py_in:
```diff
@@ -34,6 +37,25 @@ def kmeans_validate_src(schema_madlib, rel_source, **kwargs):
+                """.format(**locals()))
+            return view_name,True
+        else:
+            plpy.error(
```
How does failure of the regex mean something doesn't exist in the database?

Maybe what you want is a `try:` block around the execute that returns this error, and a separate error if the regex doesn't match? If that's the case, I suggest changing the regex to `if not p.match(...): plpy.error()`.

orhankislal commented:
Hi @decibel, thanks for your comments. I removed the regex entirely and replaced it with a built-in madlib function. I addressed your other comments as well. Please let me know what you think.

iyerr3 commented on src/ports/postgres/modules/kmeans/kmeans.py_in:
```diff
@@ -34,6 +38,34 @@ def kmeans_validate_src(schema_madlib, rel_source, **kwargs):
+    if columns_exist_in_table(rel_source, [expr_point]):
+        return rel_source, False
+
+    if is_var_valid(rel_source, expr_point):
```
An alternative to this is to use `pg_typeof` to get the type of the expression. See `get_expr_type` in `validate_args.py_in` and check for '[]' in the result. 
This allows rejecting non-array expressions. 
@decibel any concerns with using `pg_typeof` here?

orhankislal replied:
I think this is a better solution but we still allow non numeric arrays.
